### ISSUE TITLE
[snippets] Cannot override/disable snippets defined in extensions

### ISSUE THREAD

**Issue by ramya-rao-a** [microsoft/vscode#10565]
One of the Go Extension users talks about the ability to override/disable the snippets defined in the Go extension.
[vscode-go/#385](https://github.com/Microsoft/vscode-go/issues/385)

This applies to all extensions that define snippets and so adding this issue as a feature request

**Comment by zertosh** [comment]
I was just wondering this too. Seems like you can't turn off languages at all. I want to disable javascript snippets, and the language features (like suggestions, hover, etc.) from typescript, so I can run my own language service for javascript. I'd still want to keep the syntax though.

**Comment by aeschli** [comment]
IMO the best is if snippets are contributed by separate external extensions, so it's possible to uninstall them.
Note that if you don't like snippets at all you can disable them in code assist with
  `"editor.snippetSuggestions": "none"`

**Comment by Davilink** [comment]
Sometimes you don't want to shutdown all snippet suggestion, but just override some of them,  like the "cw" for C#. the default will write « System.Console.WriteLine(""); », i would like « Console.WriteLine("") », i perfectly know the the first one will work with a charm, but i just LIKE the second syntax, but when i add a snippet it's show 2 suggestion with the same prefix and that is confusing. We know, we are picky, but we like to have to possibilities to custom our editor at our taste.

**Comment by gluons** [comment]
I'm using JavaScript snippets extension and I always see 2 `if` snippet, 2 `for`, and more....
I can't type prefix and press <kbd>enter</kbd> immediately. I have to move arrow key to find snippets.
That really annoy me.

**Comment by JohnWeisz** [comment]
Also, some snippets just make _no sense at all_. For example, the `for` snippet in TypeScript creates a dummy array iteration:

```ts
for (var index = 0; index < array.length; index++) {
    var element = array[index];
    
}
```

I don't even...

 1. Why would I use a `for` loop to iterate an array in the first place? (unless I'm iterating MB-sized arrays in a worker)
 2. Why does the snippet assume I'm putting opening braces on the same line?
 3. Why does it assume I would _ever_ user `var` for block-scoped variables?
 4. Bonus: who on Earth uses `index` as a loop counter variable name?

I support the above request. _Please_, either make built-in snippets overridable, or nuke them, and let users create their own, without having to put up with useless noise continuously.

**Comment by Kitanotori** [comment]
Although currently it seems disabling particular snippets is not possible, it is possible to put your custom snippets first in the list in order to avoid skimming through the snippet menu. Just name your snippet so that the sorting algorithm puts it before the unwanted snippets (such a `1-<name of your snippet>`).

**Comment by ramya-rao-a** [comment]
Like @Kitanotori mentioned most of the scenarios listed here will be solved with the user defined snippets being sorted above the built-in snippets and the ones from extensions.

So if you don't like the `for` snippet, create your own, and that would be sorted before the other `for` snippet.
The label for the snippets will include both the name of the snippet and the prefix, making it easy to know which one is yours and which is from built-in or from the extension.

<img width="538" alt="screen shot 2017-05-03 at 6 35 48 pm" src="https://cloud.githubusercontent.com/assets/16890566/25687950/669f6984-302f-11e7-8c36-d476e5223cf0.png">

There is another scenario where this does not help. Say you have 5 javascript related extensions and all of them have the `for` snippet. Now your suggestion list just bloats up.

So it is still valuable to have a way to disable snippets not only built-in but also from extensions. We need to be able to provide the extension id and the snippet prefix and ask to disable it.

**Comment by ramya-rao-a** [comment]
My proposal would be to add a new field while snippet creation
```json
    "For Loop":{
         "prefix": "for",
         "body": "",
         "disableFromExtension": "typescript"
     }
 ```

When snippets are registered in https://github.com/Microsoft/vscode/blob/1.11.2/src/vs/workbench/parts/snippets/electron-browser/snippetsService.ts#L58 we can maintain a separate map of languageId -> snippets-to-disable

And they can be filtered out in https://github.com/Microsoft/vscode/blob/1.11.2/src/vs/workbench/parts/snippets/electron-browser/snippetsService.ts#L65

@jrieken Thoughts?

**Comment by Kitanotori** [comment]
Just noticed that identifiers (e.g. import { log } from '...') have preference over snippets, so in such case it seems to be impossible to have custom snippets come first.

**Comment by MilyMilo** [comment]
Also, it would be great to disable keywords suggestions. It basically blocks snippets because we have to use arrow down instead of just hitting enter. 
![img](https://user-images.githubusercontent.com/24303006/27806122-fe45ca9c-6038-11e7-9fec-adc0c77f9889.png)

**Comment by amackintosh** [comment]
This is incredibly annoying especially coming from Atom Editor.

I think we should be able to disable them from the snippet menu, such as a 'hide' option so you can just clean the snippet menu as it gets bad items in it. In the post above mine, imagine the 'hide' option appearing along the right side of the menu for each item.

I think this would be the most efficient and noob-user friendly solution, that would make it very intuitive to discover and clean up.

I also find it comical how you cannot set user defined snippets as the default (but, I wouldn't care about if I could simply disable them from the snippet menu).

While I am mentioning, I also find it annoying how I have to press tab twice to access the snippet. There comes a point where typing at 100WPM, its literally faster to just type it out rather than performing what is essentially some "cryptic video came cheat code input" to use a time saving device.

**Comment by ramya-rao-a** [comment]
@amackintosh What's the "tab twice" scenario you mentioned?

**Comment by amackintosh** [comment]
I added a user snippet for "log", and then after in the code: when I type "log", it displays the snippet menu, then I press tab and nothing happens, then I press tab again and it puts in <log> </log>.

I am used to Atom, using which I type "log" and then press tab and it puts console.log().

To confirm my gripe, I cannot find a way to achieve this in VSCode without doing:
type "log", press down-arrow key, press tab

I was able however, to change the name of my snippet to "1Console Log" and now it sometimes displays as the first and/or only snippet entry for "log". I haven't been able to figure out the logic yet. Sometimes it displays two logs (at which time I need to press tab twice), and other times it works as intended without having to press tab twice.

**Comment by ramya-rao-a** [comment]
@amackintosh 

> it displays the snippet menu

Do you mean the suggestion/auto completion list? 

<img width="536" alt="screen shot 2017-07-23 at 9 41 48 pm" src="https://user-images.githubusercontent.com/16890566/28508438-b8b75b00-6fef-11e7-9735-a4b2d2eeade8.png">

When no suggestion list is present, `log` followed by tab expanding to `<log></log>` is due to emmet which gets triggered on tab. You can avoid that by setting `emmet.triggerExpansionOnTab` to `false`.
We are moving away from this model of `tab` being tied to emmet, that will be out in the next update.

When you end up pressing tab twice do you get the snippet you configured or you get <log></log>?

**Comment by gluons** [comment]
@amackintosh `"editor.snippetSuggestions": "top"` may help you to get rid of arrowing down. 🙂

**Comment by amackintosh** [comment]
I get that <log> </log> snippet when I press tab twice. (also, yes I mean't suggestion/auto completion list)

Thanks @gluons, that solved my issue 100%.

**Comment by jiayihu** [comment]
Related: for some reason, in `.tsx` (Typescript React) files the built-in `import` snippet expands to `import $2 = require('$1')` instead of `import { $2 } from '$1'`. The possibility to change default snippets would allow me to override the former behaviour.

**Comment by aristotele** [comment]
I try to edit built-in snippets code, use notepadd++ in admin mode and it will work.

**Comment by LC43** [comment]
we use a specific code standard that has different spacing rules, and having to constantly change the spaces for the autocompleted code.

by being able to override the default rules, i could copy the php.json to the user snippets with the correct code standards.

+1

**Comment by fabiospampinato** [comment]
This is super annoying, I'm writing my own snippets to replace the default ones because I have a different coding style, and this is what I get:

<img width="526" alt="screen shot 2017-09-13 at 22 16 26" src="https://user-images.githubusercontent.com/1812093/30398796-f8f32374-98d1-11e7-8712-2386054d7b9b.png">

I'm not sure what the general solution should be, but if a user defines a snippet with the same prefix and description as one defined by an extension I'm pretty sure it means the one provided by the extension should be overridden.

@ramya-rao-a  Looks like there's no ETA for this issue, would you consider a PR for fixing the aforementioned issue about overriding?

**Comment by ramya-rao-a** [comment]
PRs are always welcome.
cc @jrieken

**Comment by jrieken** [comment]
Yes, PR welcome tho the idea is a little bigger than just for snippets. Users also complain about certain keybindings, or menu-contributions. So, this should be done such that it allows to disable/enable various bits of the static contributions (those that are read from package.json).

**Comment by matthewmueller** [comment]
Interestingly snippets also don't just override the exact matches, but matched suffixes. For example, an existing snippet is:

```
lp| => log.Println("|")
```

and if I try to define `;lp` as my new snippet, I'll always get No Suggestions:

```
;lp| => No Suggestions
```

Even if I define it as `glp` or anything else I continue to get "No suggestions". So it appears to match against **all** suffixes.

In my experience, snippets are a personal preference, so I'd prefer a way to either override default Go snippets or remove all of the pre-existing ones to start fresh.

**Comment by m3rabb** [comment]
Solution for MacOS: Right click on the _Visual Studio Code.app_ and select _Show Package Contents_. Drill down to the ```Contents/Resources/app/extensions/javascript/snippets``` directory, within, you will find the extensions files grouped by language. Within, I was able to find ```javascript/snippets``` and modify the existing snippets, or delete the snippet json files entirely. Restart VSC, and enjoy! :-)

**Comment by kobus1998** [comment]
I might be a little late, but I simply type the first letter of the prefix twice.

```
llog // console.log()
eeach // $1.forEach(e => {$2})
```
It will most likely always be on top.

**Comment by matthewmueller** [comment]
@kobus1998 unfortunately that doesn't seem to work if `log` were to be defined. might just be my configuration though.

**Comment by kobus1998** [comment]
@matthewmueller Hmm, weird. Could you show me your snippet? Maybe you changed the name of the snippet instead of the prefix.

```
"Console log": {
	"prefix": "llog",
	"description": "Log to console",
	"body": [
		"console.log($1)"
	]
},
```

**Comment by florinpatrascu** [comment]
on OSX, I am simply going to `~/.vscode/extensions`, chose the extension I wand to disable the snippets from and empty its `snippets/snippets.json` file. I know is barbarian and have to repeat this every time there is an update for that extension, but it is better than having the undesired snippet being offered every time. Plus ... you can always have a batch that copies an empty `snippets.json` into the right folder, if the extension gets updated too often ;) Hope the team here finds a better way to let us selectively disable the snippets from extensions. HTH

**Comment by brainsbucket** [comment]
The solution is here : https://stackoverflow.com/questions/40110541/how-to-edit-existing-vs-code-snippets
Simply add this : "editor.snippetSuggestions": "top" in your user settings file.

**Comment by florinpatrascu** [comment]
I’m afraid we’re referring to different issues. I and some of us here, were looking for a vscode solution we could use for disabling the snippets completely, from specific snippets, selectively.

**Comment by matthewmueller** [comment]
I want to summarize some issue(s) I have with the current implementation and try to provide some actionable feedback. 

I apologize if this is the wrong issue and I'm happy to post elsewhere. I've taken some time to try to reduce the number of false positives from the interplay between navigation, snippets and autocomplete. 

I'm using Go as my primary language, but I'd imagine this would apply to many languages:

**1. Autocomplete should in not mess with navigation**

I think we've all got accustomed to this from various editors, but it should not be like this. What this means is that the arrow keys, space bar and enter should be for code navigation and not for autocomplete navigation and selection. Tab can be for selecting the focused autocomplete item since I don't think there's any languages where tabs matter (mid-identifier). You can do this right now in your editor, by messing with the keybindings. Here's a fix to this:

```json5
  // disable all auto-complete triggers, except tab
  {
    "key": "down",
    "command": "-selectNextSuggestion",
    "when": "editorTextFocus && suggestWidgetMultipleSuggestions && suggestWidgetVisible"
  },
  {
    "key": "up",
    "command": "-selectPrevSuggestion",
    "when": "editorTextFocus && suggestWidgetMultipleSuggestions && suggestWidgetVisible"
  },
  {
    "key": "enter",
    "command": "-acceptSelectedSuggestionOnEnter",
    "when": "acceptSuggestionOnEnter && editorTextFocus && suggestWidgetVisible && suggestionMakesTextEdit"
  }
```

With this, you can still navigate on autocomplete by pressing cmd+arrow.

**2. Snippets are a personal preference (like themes & font)**

I don't think you should bundle snippets in the same way that you bundle language support. Maybe it's okay as a starting point, but there's gotta be a way to turn the defaults off, especially since there's some very opinionated ones single-key snippets like (e.g. `m` => `make($0, 0)`).

I do think the order of snippets should be above the scope (e.g. `"editor.snippetSuggestions": "top"`) because you want to deterministically be able to kick off a snippet, not have it dependent on what's currently in scope (something Sublime still hasn't fixed).

Since you sort of want both what's in scope and your snippets at your disposal at any given time, I tend to prefix my snippets with a `;` because it's easy to access from the home row and there's less issue with short variables (like `m := "hi"`) that are in scope. Maybe there could be a standard key like that for what ships with the standard snippet libraries.

**3. Better snippet interface**

I mentioned this before but because snippets are such a personal thing, they really need to be easier to tweak/update. Textmate had this awesome little snippet interface:

![mate](https://manual.macromates.com/images/command_editor.png)

Seems like we've been going down since then first with Sublime then Atom. I believe snippets are hugely underrated and I think part of that is because they're currently pretty painful to create and update.

---

Please don't take this feedback the wrong way, I really love the work you all are doing on this editor. I'm just a fan that wants to see VSCode get even more awesome :-D

**Comment by mickenorlen** [comment]
+1, I've also noticed many of the core snippets basically are broken eg: https://github.com/Microsoft/vscode/blob/master/extensions/php/snippets/php.snippets.json

```
	"function …": {
		"prefix": "fun",
		"body": [
			"${1:public }function ${2:FunctionName}(${3:${4:${5:Type} }$${6:var}${7: = ${8:null}}})",
			"{",
			"\t${0:# code...}",
			"}"
		],
		"description": "Function"
	},
```

It doesn't make sense to jump 5 times inside the param area. Also double dollar syntax ($$) does output single dollar ($) but also creates an extra snippet jump. Correct way to escape seems to be backslash dollar (\$). 

Would love to be able to edit snippets in a similar way to shortcuts and configs which you have already done so well. Simply need interface for viewing at least all core snippets and then right click edit -> copy and jump to snippet file where you can quickly modify it.

Thanks for all your work!

**Comment by segevfiner** [comment]
I think it would make sense for snippets with the same name to coalesce, so that if you define a snippet with the same name as one from an extension/global snippet, only the snippet with the more specific scope will appear. (I think that's the behavior in Atom).

You also need the ability to remove a snippet, I think you can define a user snippet without the "body" attribute to do so in Atom, but that's not documented anywhere. It can also be helpful to completely disable all snippets from a certain package if you don't like them at all and want to define your own.

**Comment by stamminator** [comment]
`"editor.snippetSuggestions": "top"` may work for @gluons and others, but it creates other unintended behavior that's quite disruptive. VS Code seems to give too much "weight" to snippets when this option is set to `top`, sorting it at the top of the list _no matter what_, even if there are other, more matched options in the list.

![snippet-sort-top-issue](https://user-images.githubusercontent.com/10633170/38750068-c1c1df02-3f21-11e8-8a72-f412ccdbed47.png)

In this example, the `forof` snippet, which is only a fuzzy match, is placed above `foo`, which is an _exact_ match. `foo` should be the top result. 

I'd really like to see another option for the `"editor.snippetSuggestions"` setting, maybe `"preferred"` or something like that. The user should be able to say "hey, I prefer snippets over the default language suggestions" without taking an all-or-nothing approach and throwing everything other than snippets to the bottom.

**Comment by rsxdalv** [comment]
On Windows, but presumably all platforms - going to **VS Code installation**, finding **snippets.json** under **typescript-basics** and setting it to **`{}`** will get rid of what 99% are probably in this thread for. All those **class**, **for of**, **if** and frankly useless snippets. Especially when you want to add a className but get a class inside JSX.

**Comment by leerssej** [comment]
**protostyle:**
The extensions directory is here:
  Mac/Linux:  `$HOME/.vscode/extensions/`  
  Windows: `%USERPROFILE%\.vscode\extensions/`

I selected the extension I wanted to modify, dove into the snippets directory found there (if it had one) and modified as desired.

Then I tied a string on my pinky to remember that if/when I *update* that extension someday, all my personal modifications will get overwritten/replaced out with the updated version of the file.
(unless of course I squirrelled away my changed .json file outside of the extension's directory...)

Thanks to [here][1] for helping point me towards the proper directory.
  [1]: https://www.theregister.co.uk/2015/12/16/visual_studio_code_extensions/

**Comment by Lester5123** [comment]
search "snippets" folder in:
(vscode app)/Contents/Resources/app/extensions/(language)/
or
~/.vscode/extensions/(language plug)/
edit the json file

**Comment by stamminator** [comment]
So we've seen hacks to remove or modify snippets while keeping keywords, like this...

![image](https://user-images.githubusercontent.com/10633170/41730909-01eec61c-754b-11e8-8aca-31f545f5eae9.png)

But we haven't seen anything addressing an equally important customization that I and many others would prefer: removing keywords so that snippets can be easily used, like this:

![image](https://user-images.githubusercontent.com/10633170/41730986-349f7570-754b-11e8-91b8-e31011428754.png)

So we have the _original_ issue brought up by the OP, which still needs an official solutions. Manually modifying files that can just get overwritten is not a great solution.

We also have the issue of built-in language keywords not being able to be hidden so that our hacked-up custom snippets can take the spotlight.

And we also have the issue of needing an additional `"editor.snippetSuggestions"` setting which prioritizes snippets without throwing them to the top without regard to search relevance.

Has anyone received official word on _any_ of these issues?

**Comment by Brunni132** [comment]
No progress on this?

**Comment by nealot** [comment]
I have tab complete enabled which is supposed to immediately trigger a snippet on prefix match, but because of duplicate snippets it only displays the suggestion menu.

**Comment by daraclare** [comment]
Is there an update on this? Turning off the default suggestions and keywords would be great 👍

**Comment by nin-jin** [comment]
Please, don't enforce snippets.
1. Different teams have different code styles.
2. Different developers have different preferences.
3. In different contexts developer has different intentions.
4. Unnecessary snippet makes user angry. Very angry!!!!!!!!!1111111111

Your snippets dramatically had broken my coding experience and coding performance. **It's better to make user less angry in 10% cases than make more happy in 90% cases**. Think about it. It's important UX rule.

**Comment by Santinell** [comment]
@nin-jin Fully agree with you. This is the main reason why I refused of using vscode

**Comment by rsxdalv** [comment]
I'd like to ask people to stop commenting when a +1 would have sufficed. I understand that you feel the need to express your frustration as you've just stumbled upon this issue and might experience a slight shock. But the fact is that for those who have discovered it before you, a hopeful e-mail with a solution is yet another comment that contributes nothing but anger.

The fact of the matter is, there are higher priority issues, and there are perfectly sound workarounds for this. If they have proved too difficult to do, ask for help, someone might, and they'll become better documented overall.

Right now you are merely pushing people away from subscribing to this thread, so that if a dev came here and said: "we won't do this", the people who cared would not be notified, and would not be able to react.

**Comment by NanerLee** [comment]
It seems that Atom has solved a similar problem -- [Disabling snippets included from packages](https://github.com/atom/snippets/issues/55#event-1910064756). 

Can it provide some advice on vscode?

**Comment by svipben** [comment]
Personally I care mostly about built-in snippets. It will be very nice if we had an ability to delete or edit them. Thanks!

**Comment by bakape** [comment]
What I currently do is run `find ~/.vscode -name snippets -exec rm -r {}
\;` each time and extension updates. It's a dumb hack but it works, if
anyone needs it.

**Comment by cra0zy** [comment]
PS. `editor.snippetSuggestions: "none"` if anyone needs it.

**Comment by MarkKahn** [comment]
Please fix this.  This issue was opened 2.5 years ago and I've been following it for at least half of that time.  Just give us a `editor.disableDefaultSnippets: true` or similar that turns off all default and extension snippets and just leaves specific ones created by the user.

**Comment by gluons** [comment]
I also need API for snippets extension.

**Comment by svipas** [comment]
@jrieken What do you think about `"editor.suggest.defaultSnippets": "on" | "off"`? I can try to create PR.

**Comment by bakape** [comment]
Only not suggesting them is half the problem. They would still autocomplete
on keyword + tab (assuming you have this behaviour enabled).

**Comment by Managor** [comment]
I'm bumping this thread as I've recently grown annoyed with the the fact that the c/c++ snippets are bound within a binary file and there's no way to override or edit them. I prefer not to litter my suggestion menu with duplicates.

**Comment by sleighsoft** [comment]
Would love to see this: https://github.com/microsoft/vscode/issues/10565#issuecomment-472358909

**Comment by GeorgeXc** [comment]
> IMO the best is if snippets are contributed by separate external extensions, so it's possible to uninstall them.
> Note that if you don't like snippets at all you can disable them in code assist with
> `"editor.snippetSuggestions": "none"`

It works

**Comment by incleaf** [comment]
I used to use snippets a lot when I was using WebStorm. 
I started not using snippets on VS Code because default snippets are not useful for me and I can't even disable it.

**Comment by kevin-ashton** [comment]
For what it's worth when it comes to the annoying build in snippets for typescript they are easy enough to snipe after each release. I assume it would be something similar for other languages. I run the following script every time they begin showing again. 

```bash
#  This is in my ~/.profile
function vsCodeClearSnippits() {
    echo "Reseting Snippits"

    # Typescript
    rm /Applications/Visual\ Studio\ Code.app/Contents/Resources/app/extensions/typescript-basics/snippets/typescript.json
    echo {} > /Applications/Visual\ Studio\ Code.app/Contents/Resources/app/extensions/typescript-basics/snippets/typescript.json

    # Javascript
    rm /Applications/Visual\ Studio\ Code.app/Contents/Resources/app/extensions/javascript/snippets/javascript.json
    echo {} > /Applications/Visual\ Studio\ Code.app/Contents/Resources/app/extensions/javascript/snippets/javascript.json
}
```

![image](https://user-images.githubusercontent.com/1718984/58737132-a6ae2900-83bd-11e9-9a63-8bfc68a7193e.png)


Anyway this is hardly the ideal solution because it does revert occasionally and I have to run the `vsCodeClearSnippits` from the terminal again. I assume vsCode does this when it updates or detects it has been changed.  Anyway annoying to have to run the command again but it works well enough. 

Hopefully they provide an actual solution in the future.

**Comment by Siiiimon** [comment]
has there been an update on this? (apart from editing extensions in a hacky way)

**Comment by adamreisnz** [comment]
Seems like people have suggested a very sensible setting in the form of `editor.suggest.disableDefaultSnippets": true | false` with the offer of a PR several months ago. 

Anyone for the core team care to comment on that suggested approach and green light it?

**Comment by t7yang** [comment]
The default built-in snippets (like for JS /  TS) is pretty annoying.
At least make the user custom snippets and user installed snippets before the default.

**Comment by fabriciomsantos** [comment]
Any news about override the keyword if has a snippet with the same prefix name?
to resolve this: 
![Capturar](https://user-images.githubusercontent.com/37518637/61330853-f6727380-a7f6-11e9-91f0-61044ffae884.PNG)

**Comment by inwardmovement** [comment]
Any news on providing a setting to disable extension-specific snippets?

![image](https://user-images.githubusercontent.com/9438102/61698795-fe23a200-ad39-11e9-83b0-99e03153a572.png)

I don't want to change the order of suggestions — I still want my user-defined snippets to show up first — but only disable unwanted extensions snippets...

**Comment by svipas** [comment]
@kevin-ashton Maybe you or somebody else know where are snippets in Windows?

I'm creating an extension which will remove all global snippets. I think of two ways, first, one removes all global snippets on VS Code startup and the second one have a command to remove all global snippets. Should I implement both of them or only command to remove all global snippets?

**Comment by dotiful** [comment]
@svipas hey, how is it going, any progress?
sounds like a good solution until native support appears

**Comment by svipas** [comment]
@dotiful Hi, if I will have time today I will publish this extension. It would have picker with all global snippets from builtin extensions and downloaded, so you could choose which one to delete or all of them.

**Comment by gluons** [comment]
@svipas Does your extension have **disabling snippets** feature instead of delete?  
So, we'll able to bring it back when we need it. Or show/hide as we wish.

**Comment by svipas** [comment]
@gluons It's way easier to delete them, but I will see if I can add the ability to disable them instead of deleting them.

**Comment by Vortex93** [comment]
@svipas Any update on the extension that you are working on?

**Comment by svipas** [comment]
I did some tests and seems that we can't delete snippets instead I will disable them, so you could enable/disable them from picker. After new VS Code release 1.37.0 I will publish extension since VS Code provides `fs` wrapper so it would be very easy to fully use VS Code API. Stay tuned I will update here about extension which will be published probably this week 🚀

**Comment by svipas** [comment]
Extension is already created and works amazing! You can disable or enable snippets from any extensions, built-in or manually installed. Extension doesn't delete any files!

Repository: https://github.com/svipas/vscode-control-snippets
VS Code extension: https://marketplace.visualstudio.com/items?itemName=svipas.control-snippets

How it looks like:
![](https://raw.githubusercontent.com/svipas/vscode-control-snippets/master/control-snippets.png)

**Comment by danielo515** [comment]
So this is a very important feature that annoys developers every single day of their lives.
I'ts have been 3 years and several duplicates since it was first reported, and I didn't saw any answer from the core team. I think we are doomed.

**Comment by leerssej** [comment]
Thank goodness they have open sourced it so we could get something done
about it!
Thank you very much, Benas!!!

**Comment by escape0707** [comment]
> So if you don't like the `for` snippet, create your own, and that would be sorted before the other `for` snippet.

I'm wondering what did I miss to get the [described](https://github.com/microsoft/vscode/issues/10565#issuecomment-299080618) behavior?

Screenshots:
![image](https://user-images.githubusercontent.com/11014745/68384854-7a2fe180-0193-11ea-9e7a-150967df4433.png)
![image](https://user-images.githubusercontent.com/11014745/68384871-82881c80-0193-11ea-9ebc-c0554e5d2f2b.png)

**Comment by yevagorbachev** [comment]
Rename the snippet files from "`language.json`" to "`language.disabled`". They are located in `%USERPROFILE%\.vscode\extensions`

**Comment by ioquatix** [comment]
I found the snippet and added:

```
'.ruby.source':
  'Add ‘# =>’ Marker':
    'prefix': '#'
```

to my `~/.atom/snippets.cson` and it stopped the snippet from firing.

**Comment by ryuta69** [comment]
All I want is, 

- Snippet has an option to hide
- There will be three dots in suggestion window to control to hide or not.

So I can choose good one left, bad one hidden.

**Comment by VijayaKatikireddy** [comment]
This will be a great feature,  and helps avoid overloading.

**Comment by antonmyrberg** [comment]
The plugin wasn't was I was looking for, I didn't want to disable all snippets for a lang, just overwrite the existing one.

What I ended up with was just prefixing all my snippets like `>ue` and `>us` (useEffect and useState). 
Never collides with the built-ins, but I would really prefer to be able to declare snippets like @ramya-rao-a suggested using `"disableFromExtension": "typescript"` or similar.

**Comment by siimsepp** [comment]
I had the same problem of annoying default snippets. I found a file named javascript.code-snippets there:
C:\Users\<USERNAME>\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\javascript\snippets
After removing it and restarting VS Code I see only my own snippets. If you only want to remove some default snippets then you can edit that file. Simply cut out annoying for-loop snippet and only that one is gone. This may be a temporary fix until the next update but it is easy to repeat if you keep your own version of the file and replace the default one with it after the update.

**Comment by connesy** [comment]
On linux, the default code snippets can be found in `/usr/share/code/resources/app/extensions/[language]/snippets/[language].code-snippets`, so for JavaScript that would be `/usr/share/code/resources/app/extensions/javascript/snippets/javascript.code-snippets`. 
As with [siimsepp's answer](https://github.com/Microsoft/vscode/issues/10565#issuecomment-703070378), these changes take effect after restarting VSCode.

**Comment by Pablion** [comment]
these are nice workarounds but not the ultimate solution.

**Comment by novachief101** [comment]
This issue has been open since 2016, seriously, my favovurite editor? All of the options are just _work arounds_, I really do hope that someone does something about this. 😞

**Comment by bakape** [comment]
I took a stab at it a few years ago. It's non-trivial, unless you already
know your way around the snippet codebase.

**Comment by novachief101** [comment]
I know it's pretty easy but hey, it wouldn't be too hard to directly implement it.

**Comment by AllanDaemon** [comment]
There is a major extension that provides a snippet with the prefix `.`, so every time I use a `.` it changes to `self.`.

This is so bad that many times I wast more time undoing this than doing the task I supposed to. The fact that I can't type `.` is pretty annoying and this has 4 years is frustrating.

**Comment by jrieken** [comment]
This will happen for 1.52 and this enabling/disabling snippets works from the F1 > Insert Snippet list

* Select F1 > Insert Snippet
* find an extension snippet that you want to hide
* select the eye-icon to hide/show it in IntelliSense

Note that hidden snippets are synced across devices when using setting-sync, so this needs to be done only once per extension snippet. 

![Nov-04-2020 17-21-42](https://user-images.githubusercontent.com/1794099/98137445-4a104100-1ec2-11eb-9c61-a530217cf137.gif)

**Comment by danielo515** [comment]
@jrieken does that also apply to built in extensions? Those are the ones that I find more annoying (all those old-style snippets like for loops, selfs, this, etc)

**Comment by jrieken** [comment]
@danielo515 Yes - it will apply to all snippets from extensions - builtin or user-extensions. 

This will be in next `1.52` insiders: https://code.visualstudio.com/insiders/

**Comment by jrieken** [comment]
To verify

* make sure that you can hide/show _extension_ snippets, not user nor workspace snippets, via "F1 > Insert Snippet"
* make sure that hidden snippets do not show in IntelliSense
* make sure that hidden snippets still show in "F1 > Insert Snippet"

**Comment by novachief101** [comment]
spam, or gmail thinks so
